### ISSUE TITLE
Segmentation fault or no convergence when using XLA_CPU with distributed TensorFlow

### ISSUE THREAD

**Issue by sfvaroglu** [tensorflow/tensorflow#17458]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: I have modified tensorflow/examples/tutorials/mnist/mnist_softmax_xla.py to use MonitoredTrainingSession and PS/worker hosts.
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: source
- **TensorFlow version**: 1.6.0
- **Python version**: 2.7
- **Bazel version (if compiling from source)**: 0.11.0
- **GCC/Compiler version (if compiling from source)**: 5.4.0
- **CUDA/cuDNN version**: n/a
- **GPU model and memory**: n/a
- **Exact command to reproduce**: run with a local PS and a local worker

### Describe the problem
I'm seeing segmentation fault or no convergence when using mnist_softmax_xla.py in a distributed environment (one PS and one worker, both local) and `train_step` placed on XLA_CPU.

```
with tf.device('/device:XLA_CPU:0'):
  train_step = tf.train.GradientDescentOptimizer(0.5).minimize(cross_entropy)
```

Is this the expected behavior? What is the recommended way of using XLA_CPU in distributed training?

**Comment by asimshankar** [comment]
Segmentation faults are rarely expected behavior :)
Can you share code to reproduce the problem? Ideally, as small a code snippet as possible that reproduces it.

More information would also be helpful, such as:
- Details of the segmentation fault (do you have a stack trace for example?)
- Where is the segmentation fault occurring - on all workers? on the PS? etc.

(CC @tatianashp )

**Comment by sfvaroglu** [comment]
The behavior is inconsistent. What I observed:
- segmentation fault on worker side (the number of steps where segmentation fault occurs is inconsistent), or
- no segmentation fault, but accuracy < 0.1

The same script runs fine (i.e. no segmentation fault or convergence problem) if `with tf.device('/device:CPU:0')` is used. 

```
def main(_):
  ps_hosts = FLAGS.ps_hosts.split(",")
  worker_hosts = FLAGS.worker_hosts.split(",")

  cluster = tf.train.ClusterSpec({"ps": ps_hosts, "worker": worker_hosts})

  server = tf.train.Server(cluster,
                           job_name=FLAGS.job_name,
                           task_index=FLAGS.task_index)


  if FLAGS.job_name == "ps":
    server.join()
  elif FLAGS.job_name == "worker":
    global_step = tf.train.get_or_create_global_step()
    
    hooks=[tf.train.StopAtStepHook(num_steps=10)]

    mnist = input_data.read_data_sets(FLAGS.data_dir)

    with tf.device(tf.train.replica_device_setter(cluster=cluster)):
      # Create the model
      x = tf.placeholder(tf.float32, [None, 784])
      w = tf.Variable(tf.zeros([784, 10]))
      b = tf.Variable(tf.zeros([10]))
      y = tf.matmul(x, w) + b

      # Define loss and optimizer
      y_ = tf.placeholder(tf.float32, [None, 10])

      # The raw formulation of cross-entropy,
      #
      #   tf.reduce_mean(-tf.reduce_sum(y_ * tf.log(tf.nn.softmax(y)),
      #                                 reduction_indices=[1]))
      #
      # can be numerically unstable.
      #
      # So here we use tf.nn.softmax_cross_entropy_with_logits on the raw
      # outputs of 'y', and then average across the batch.
      cross_entropy = tf.reduce_mean(
        tf.nn.softmax_cross_entropy_with_logits(labels=y_, logits=y))

      with tf.device('/device:XLA_CPU:0'):
        train_step = tf.train.GradientDescentOptimizer(0.5).minimize(cross_entropy)

      correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(y_, 1))
      accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))

    config = tf.ConfigProto(allow_soft_placement=True, log_device_placement=True)
    jit_level = 0
    if FLAGS.xla:
      # Turns on XLA JIT compilation.
      jit_level = tf.OptimizerOptions.ON_1

    config.graph_options.optimizer_options.global_jit_level = jit_level

    with tf.train.MonitoredTrainingSession(master=server.target,
                                           is_chief=(FLAGS.task_index==0),
                                           hooks=hooks,
                                           config=config) as sess:

      train_loops = FLAGS.train_loop_count
      start = time.time()
      for i in range(train_loops):
        batch_xs, batch_ys = mnist.train.next_batch(100)
        (_, loss) = sess.run(
          [train_step, cross_entropy], feed_dict={
              x: batch_xs,
              y_: batch_ys
          })
        print("step = %d, loss = %f" % (i, loss))
      end = time.time()
      elapsed = (end - start)
      print("Time elapsed: ", elapsed, " seconds" )

      # Test trained model
      if i == train_loops-1:
        print("Accuracy: ", sess.run(
          accuracy, feed_dict={
            x: mnist.test.images,
            y_: mnist.test.labels
          }))
```

**Comment by sfvaroglu** [comment]
Could you elaborate on _`device:XLA_CPU:0’ is used primarily for testing?_ Using it in a non-distributed environment does not cause segmentation fault or non-convergence. Does it result in undefined behavior in distributed runtime? 

Setting config or scope to invoke XLA (as mentioned in https://www.tensorflow.org/performance/xla/jit) does not place ops on XLA_CPU device. The logs show all ops are either on `/job:worker/replica:0/task:0/device:CPU:0` or `/job:ps/replica:0/task:0/device:CPU:0`. 

Also, using `worker_device=device:XLA_CPU` in tf.train.replica_device_setter causes segmentation fault or NaN values.

What is the recommended way of using XLA_CPU in a distributed environment?

**Comment by tatianashp** [comment]
@andydavis1 Could you comment on using XLA_CPU in a distributed environment?

**Comment by andydavis1** [comment]
You should configure XLA using the pointers that tatianashp@ sent you (i.e. do not use XLA_CPU device). Configured with sesion config or jit_scope,, there is no XLA_CPU device, but XLA CPU backend will be enabled on the host (i.e. CPU:0 not XLA_CPU:0).

**Comment by sfvaroglu** [comment]
How do I verify if XLA is invoked or see which ops are compiled via XLA? I only see `LaunchOpHasKernelForDevice kernel_class_name: XlaLocalLaunchOp` in the log.

Also, I was expecting to see send/recv ops between `/job:ps/replica:0/task:0/device:CPU:0` and `/job:worker/replica:0/task:0/device:XLA_CPU:0`, but that's not the case.

**Comment by andydavis1** [comment]
There are timeline and graphviz instructions here: https://www.tensorflow.org/performance/xla/jit
XLA CPU does not currently support send/recv instructions.

**Comment by sfvaroglu** [comment]
> XLA CPU does not currently support send/recv instructions

Could you elaborate on that? I see XLA send/recv in tensorflow/compiler/tf2xla/ops/sendrecv_ops.cc. Is cross-device XLA send/recv not supported? If so, when is it planned to be supported?

**Comment by andydavis1** [comment]
Send/Recv instructions are not currently implemented on the XLA CPU backend. If you want to do distributed training and use XLA CPU, you can have TF do the send/recv between machines and pass these tensors to the TF sub-graph that you want to compile for XLA CPU. So you would just want to JIT compile the sub-set of the TF graph that runs on each machine, and use TF to do the send/recv between machines.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tatianashp** [comment]
@sfvaroglu Please reopen if you have further concerns.
